app/src/org/commcare/activities/connect/ConnectIdActivity.java

```diff
@@ -0,0 +1,166 @@
+                } else if (user.shouldForcePassword()) {
+                    navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionToConnectidPassword(ConnectConstants.CONNECT_UNLOCK_PASSWORD, user.getPrimaryPhone(), user.getPassword());
+                } else {
+                    navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionToConnectidBiometricConfig(user.getPrimaryPhone(), ConnectConstants.CONNECT_UNLOCK_BIOMETRIC).setAllowPassword(true);
+                }
+            }
+            default -> {
+                //Error, should never get here
+            }
+        }
+
+        if (requestCode != ConnectConstants.CONNECT_NO_ACTIVITY) {
+            if (navDirections != null) {
+                controller.navigate(navDirections);
+            }
+        }else{
+            navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionSelf();
+            controller.navigate(navDirections);
+        }
+    }
+
+    public static void unlockConnect(Context parent) {
+        forgotPassword = false;
+        forgotPin = false;
+
+        ConnectUserRecord user = ConnectDatabaseHelper.getUser(parent);
+        navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionToConnectidBiometricConfig(user.getPrimaryPhone(), ConnectConstants.CONNECT_UNLOCK_BIOMETRIC).setAllowPassword(true);
+        if (user.shouldForcePin()) {
+            navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionToConnectidPin(ConnectConstants.CONNECT_UNLOCK_PIN, user.getPrimaryPhone(), user.getPassword());
+        } else if (user.shouldForcePassword()) {
+            navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionToConnectidPassword(ConnectConstants.CONNECT_UNLOCK_PASSWORD, user.getPrimaryPhone(), user.getPassword());
+        }
+
+        if (navDirections != null) {
+            controller.navigate(navDirections);
+        }
+    }
+
+    public static void beginSecondaryPhoneVerification(Context parent) {
+        navDirections = org.commcare.fragments.connectId.ConnectIdRecoveryDecisionFragmentDirections.actionConnectidRecoveryDecisionToConnectidMessage(parent.getString(R.string.connect_recovery_alt_title), parent.getString(R.string.connect_recovery_alt_message), ConnectConstants.CONNECT_VERIFY_ALT_PHONE_MESSAGE, parent.getString(R.string.connect_password_fail_button), parent.getString(R.string.connect_recovery_alt_change_button));
```

Comment:
Add some newlines to break this line up and make it more readable, like before each call to parent.getString()